### PR TITLE
kernel/binary_manager : Use logic to check checksum with crc

### PR DESCRIPTION
**Pull request by YiHyunJin**
The binary manager should check the checksum before loading.

### CODE REVIEW

File: os/kernel/binary_manager/binary_manager_load.c [hunk 1]
```diff
@@ -55,6 +55,37 @@ typedef struct binary_header_s binary_header_t;
...
+	uint8_t buffer[CRC_BUFFER_SIZE];
+
+	ret = lseek(fd, CHECKSUM_SIZE, 0);
+	if (ret < 0) {
+		bmdbg("Failed to lseek : %d, errno %d\n", devname, ret, errno);
+	}
+
+	while (file_size > 0) {
+		if (file_size > CRC_BUFFER_SIZE) {
+			ret = read(fd, (FAR uint8_t *)buffer, CRC_BUFFER_SIZE);
```
Comment (jeongchanKim) [inline os/kernel/binary_manager/binary_manager_load.c]: You'd better check the ret value for error case.

File: os/kernel/binary_manager/binary_manager_load.c [hunk 2]
```diff
@@ -55,6 +55,37 @@ typedef struct binary_header_s binary_header_t;
...
+		} else {
+			ret = read(fd, (FAR uint8_t *)buffer, file_size);
+		}
+
+		check_crc = crc32part(buffer, ret, check_crc);
```
Comment (jeongchanKim) [inline os/kernel/binary_manager/binary_manager_load.c]: Please add `#include <crc32.h>`.

File: os/kernel/binary_manager/binary_manager_load.c [hunk 3]
```diff
@@ -91,17 +124,32 @@ static int binary_manager_read_header(int bin_idx, int part_idx, binary_header_t
 		return ERROR;
 	}
 
+	/* Read the checksum */
+	ret = read(fd, (FAR uint8_t *)&crc_hash, sizeof(uint32_t));
```
Comment (jeongarmy) [inline os/kernel/binary_manager/binary_manager_load.c]: need to check ret value

**Comment by seinfra** [comment]
Target : [da02b3b9c8e012ed44112d57bad8e24505566d60] - Code Rule Check (C++) OK.

**Comment by seinfra** [comment]
Target : [da02b3b9c8e012ed44112d57bad8e24505566d60] - Code Rule Check OK.

File: os/kernel/binary_manager/binary_manager_load.c [hunk 4]
```diff
@@ -55,6 +55,37 @@ typedef struct binary_header_s binary_header_t;
...
  * Private Functions
  ****************************************************************************/
+
+static int binary_manager_verify_checksum(int fd, int file_size, uint32_t crc_hash)
+{
+	int ret;
+	uint32_t check_crc = 0;
+	uint8_t buffer[CRC_BUFFER_SIZE];
+
+	ret = lseek(fd, CHECKSUM_SIZE, 0);
```
Comment (jeongarmy) [inline os/kernel/binary_manager/binary_manager_load.c]: <unistd.h> is needed for lseek

File: os/kernel/binary_manager/binary_manager_load.c [hunk 5]
```diff
@@ -55,6 +55,37 @@ typedef struct binary_header_s binary_header_t;
...
+	if (ret < 0) {
+		bmdbg("Failed to lseek : %d, errno %d\n", devname, ret, errno);
```
Comment (jeongarmy) [inline os/kernel/binary_manager/binary_manager_load.c]: return ERROR

**Comment by seinfra** [comment]
Target : [e5fd5ff9ac883147a28c777c0d442ab665357221] - Code Rule Check OK.

**Comment by seinfra** [comment]
Target : [e5fd5ff9ac883147a28c777c0d442ab665357221] - Code Rule Check (C++) OK.

File: os/kernel/binary_manager/binary_manager_load.c [hunk 6]
```diff
@@ -91,17 +132,32 @@ static int binary_manager_read_header(int bin_idx, int part_idx, binary_header_t
...
+	file_size = header_data->header_size + header_data->bin_size;
+
+	if (binary_manager_verify_checksum(fd, file_size, crc_hash) == ERROR) {
+		close(fd);
+		return ERROR;
+	}
+
 	close(fd);
 
 	if (ret != sizeof(binary_header_t)) {
```
Comment (jeongarmy) [inline os/kernel/binary_manager/binary_manager_load.c]: We need to check whether reading succeed first before verifying header data.
So this should be line 139

File: os/kernel/binary_manager/binary_manager_load.c [hunk 7]
```diff
@@ -55,6 +57,43 @@ typedef struct binary_header_s binary_header_t;
 /****************************************************************************
  * Private Functions
  ****************************************************************************/
+
+static int binary_manager_verify_checksum(int fd, int file_size, uint32_t crc_hash)
+{
+	int ret;
```
Comment (sunghan-chang) [inline os/kernel/binary_manager/binary_manager_load.c]: Let's rename `read_size`.

File: os/kernel/binary_manager/binary_manager_load.c [hunk 8]
```diff
@@ -55,6 +57,43 @@ typedef struct binary_header_s binary_header_t;
...
+	uint32_t check_crc = 0;
+	uint8_t buffer[CRC_BUFFER_SIZE];
```
Comment (sunghan-chang) [inline os/kernel/binary_manager/binary_manager_load.c]: crc_buffer

**Comment by seinfra** [comment]
Target : [f4324d4a35de5a307fbba40cffc668a7b9ce62fc] - Code Rule Check (C++) OK.

**Comment by seinfra** [comment]
Target : [f4324d4a35de5a307fbba40cffc668a7b9ce62fc] - Code Rule Check OK.

**Comment by seinfra** [comment]
Target : [655bdc6b82b3d015c7c0eedfa27975ab41611e01] - Code Rule Check (C++) OK.

**Comment by seinfra** [comment]
Target : [655bdc6b82b3d015c7c0eedfa27975ab41611e01] - Code Rule Check OK.

File: os/kernel/binary_manager/binary_manager_load.c [hunk 9]
```diff
@@ -91,17 +131,36 @@ static int binary_manager_read_header(int bin_idx, int part_idx, binary_header_t
...
+	/* Read the checksum */
+	ret = read(fd, (FAR uint8_t *)&crc_hash, CHECKSUM_SIZE);
+	if (ret != CHECKSUM_SIZE) {
+		bmdbg("Failed to read %s: %d, errno %d\n", devname, ret, errno);
+		return ERROR;
+	}
+
 	/* Read the binary header */
 	ret = read(fd, (FAR uint8_t *)header_data, sizeof(binary_header_t));
-	close(fd);
```
Comment (sunghan-chang) [inline os/kernel/binary_manager/binary_manager_load.c]: When `ret` is not same as `sizeof(bin_..._t)`, `close(fd)` is missing.

File: os/kernel/binary_manager/binary_manager_load.c [hunk 10]
```diff
@@ -91,17 +131,36 @@ static int binary_manager_read_header(int bin_idx, int part_idx, binary_header_t
...
-	return binary_manager_verify_header(header_data);
+	if (binary_manager_verify_header(header_data) == ERROR) {
+		close(fd);
+		return ERROR;
+	}
+
+	file_size = header_data->header_size + header_data->bin_size;
+
+	if (binary_manager_verify_checksum(fd, file_size, crc_hash) == ERROR) {
+		close(fd);
```
Comment (sunghan-chang) [inline os/kernel/binary_manager/binary_manager_load.c]: Don't you need to add debug message like other failure cases?

**Comment by seinfra** [comment]
Target : [e83a9454ddd0ffb1b3613f6bf9a654ae93024cf6] - Code Rule Check (C++) OK.

**Comment by seinfra** [comment]
Target : [e83a9454ddd0ffb1b3613f6bf9a654ae93024cf6] - Code Rule Check OK.

**Pull request merged**